### PR TITLE
Improvements to 3PE lookup API

### PR DESCRIPTION
**Pull request by leonerd**
* Merge the `instances` lists from multiple AS responses into a single combined result to the client
 * Add a `GET .../protocol/:protocol` endpoint for efficiency

Tested in https://github.com/matrix-org/sytest/pull/305

### CODE REVIEW

File: synapse/handlers/appservice.py [hunk 1]
```diff
@@ -176,12 +176,45 @@ def query_3pe(self, kind, protocol, fields):
...
+                    continue
+                if not isinstance(info["instances"], list):
+                    continue
+
+                if p not in protocols:
+                    protocols[p] = []
+                protocols[p].append(info)
+
+        def _merge_instances(infos):
+            if len(infos) == 0:
```
Comment (erikjohnston) [inline synapse/handlers/appservice.py]: Generally you'd just do `if not infos:`

File: synapse/handlers/appservice.py [hunk 2]
```diff
@@ -176,12 +176,45 @@ def query_3pe(self, kind, protocol, fields):
...
+                if only_protocol is not None and p != only_protocol:
+                    continue
+
+                info = yield self.appservice_api.get_3pe_protocol(s, p)
+
+                # Ignore any result that doesn't contain an "instances" list
+                if "instances" not in info:
+                    continue
+                if not isinstance(info["instances"], list):
+                    continue
```
Comment (erikjohnston) [inline synapse/handlers/appservice.py]: Generally you don't do type checking like this. If you're worried that `info["instances"]` isn't going to be an iterable then you can catch the `TypeError` below where you do `list(info["instances"])`. I'm also surprised this doesn't happen in `appservice_api.get_3pe_protocol`.

Unless this is acceptable behaviour for an app service I'd also log

File: synapse/handlers/appservice.py [hunk 3]
```diff
@@ -176,12 +176,45 @@ def query_3pe(self, kind, protocol, fields):
...
+                    protocols[p] = []
+                protocols[p].append(info)
+
+        def _merge_instances(infos):
+            if len(infos) == 0:
+                return {}
+
+            # Merge the 'instances' lists of multiple results, but just take
+            # the other fields from the first as they ought to be identical
+            # deep-clone the result so as not to corrupt the cached one
```
Comment (erikjohnston) [inline synapse/handlers/appservice.py]: (At not point here are we actually *deep* cloning)

File: synapse/handlers/appservice.py (hunk 1 above)
Comment (leonerd) [inline synapse/handlers/appservice.py]: Done f25d74f

File: synapse/handlers/appservice.py (hunk 3 above)
Comment (leonerd) [inline synapse/handlers/appservice.py]: Adjusted wording f25d74f

File: synapse/handlers/appservice.py (hunk 2 above)
Comment (leonerd) [inline synapse/handlers/appservice.py]: Have moved type-testing code into AS API ed44c47

Comment (leonerd) [inline synapse/handlers/appservice.py]: Also log if rejecting  776594f

**Comment by erikjohnston** [comment]
LGTM

**Pull request merged**